pkg/jobstore/boltdb/store.go

```diff
@@ -0,0 +1,1013 @@
+	return b.appendJobHistory(tx, jobState, previousState, request.Comment)
+}
+
+func (b *BoltJobStore) appendJobHistory(tx *bolt.Tx, updateJob model.JobState, previousState model.JobStateType, comment string) error {
+	historyEntry := model.JobHistory{
+		Type:  model.JobHistoryTypeJobLevel,
+		JobID: updateJob.JobID,
+		JobState: &model.StateChange[model.JobStateType]{
+			Previous: previousState,
+			New:      updateJob.State,
+		},
+		NewVersion: updateJob.Version,
+		Comment:    comment,
+		Time:       updateJob.UpdateTime,
+	}
+	data, err := json.Marshal(historyEntry)
+	if err != nil {
+		return err
+	}
+
+	// Get the history bucket for this job ID, which involves potentially
+	// creating the bucket (jobs.history.JOBID)
+	if bkt, err := NewBucketPath(BucketPathJobHistory, updateJob.JobID).Get(tx, true); err != nil {
+		return err
+	} else {
+		seq := BucketSequenceString(tx, bkt)
+		if err = bkt.Put([]byte(seq), data); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// CreateExecution creates a record of a new execution
+func (b *BoltJobStore) CreateExecution(ctx context.Context, execution model.ExecutionState) error {
+	return b.database.Update(func(tx *bolt.Tx) (err error) {
+		return b.createExecution(tx, execution)
+	})
+}
+
+func (b *BoltJobStore) createExecution(tx *bolt.Tx, execution model.ExecutionState) error {
+	if !b.jobExists(tx, execution.JobID) {
+		return jobstore.NewErrJobNotFound(execution.JobID)
+	}
+
+	execID := []byte(execution.ID().String())
+
+	if execution.CreateTime.IsZero() {
+		execution.CreateTime = b.clock.Now().UTC()
+	}
+	if execution.UpdateTime.IsZero() {
+		execution.UpdateTime = execution.CreateTime
+	}
+	if execution.Version == 0 {
+		execution.Version = 1
+	}
+
+	// Check for the existence of this ID and if it doesn't already exist, then create
+	// it
+	if bucket, err := NewBucketPath(BucketPathExecutions).Get(tx, false); err != nil {
+		return err
+	} else {
+		_, err := b.getExecution(tx, execution.ID())
+		if err == nil {
+			return jobstore.NewErrExecutionAlreadyExists(execution.ID())
+		}
+
+		if data, err := json.Marshal(execution); err != nil {
+			return err
+		} else {
+			err = bucket.Put(execID, data)
+			if err != nil {
+				return err
+			}
+
+			b.triggerEvent(jobstore.ExecutionWatcher, jobstore.CreateEvent, data)
+		}
+	}
+
+	return b.appendExecutionHistory(tx, execution, model.ExecutionStateNew, "")
+}
+
+// UpdateExecution updates the state of a single execution by loading from storage,
+// updating and then writing back in a single transaction
+func (b *BoltJobStore) UpdateExecution(ctx context.Context, request jobstore.UpdateExecutionRequest) error {
+	return b.database.Update(func(tx *bolt.Tx) (err error) {
+		return b.updateExecution(tx, request)
+	})
+}
+
+func (b *BoltJobStore) updateExecution(tx *bolt.Tx, request jobstore.UpdateExecutionRequest) error {
+	if !b.jobExists(tx, request.ExecutionID.JobID) {
+		return jobstore.NewErrJobNotFound(request.ExecutionID.JobID)
+	}
+
```

Comment:
when updating an execution, we should just care if the execution exists or not, but not the job. I understand it is a bug where an execution remains but its job was deleted or never created, but this check won't help much